src/cantools/database/utils.py

```diff
@@ -123,15 +123,31 @@ def decode_data(data: bytes,
                 decode_choices: bool,
                 scaling: bool,
                 allow_truncated: bool,
+                allow_excess: bool,
                 ) -> SignalDictType:
 
     actual_length = len(data)
-    if allow_truncated and actual_length < expected_length:
-        data = data.ljust(expected_length, b"\xFF")
+
+    if actual_length < expected_length:
+        if not allow_truncated:
+            raise ValueError(f"Wrong data size: {actual_length} instead of "
+                             f"{expected_length} bytes")
+        else:
+            # pad the data with 0xff to prevent the codec from raising an
+            # exception
+            data = data.ljust(expected_length, b"\xFF")
+
+    if actual_length > expected_length:
+        if not allow_excess:
+            raise ValueError(f"Wrong data size: {actual_length} instead of "
+                             f"{expected_length} bytes")
+        else:
+            # trim the payload data to match the expected size
+            data = data[:expected_length]
 
     unpacked = {
-        **formats.big_endian.unpack(data),
-        **formats.little_endian.unpack(data[::-1]),
+        **formats.big_endian.unpack(bytes(data)),
```

Comment:
because for whatever reason `data` now sometimes is a `bytearray` here and thus test_autosar.py fails
```
> python3 -m pytest -x --tb=native tests/test_autosar.py
[...]
  File "src/cantools/database/utils.py", line 149, in decode_data
    **formats.big_endian.unpack(data),
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: expected bytes, bytearray found
---------------------------------------------
```

I have not invested too much time in finding out why that is, or why `test_autosar` is so picky. (I suspect the latter to be `bitstruct`'s fault.) Stay tuned...

Reply:
okay, the culprit is `bitstruct.c`: it does not accept `bytearray` objects for `CompiledFormatDict.unpack`. (`bitstruct` works fine...)

Reply:
Which version of bitstruct? It should work https://github.com/eerimoq/bitstruct/pull/31/files 🤔

Reply:
`8.17.0`. (Stupid me; I had a clone of the repo pointing at this old version lying around in my PYTHONPATH.) Updating fixed it! Nice, thanks for the fix!